grafana/mimir
# Helm: run under Restricted pod security policy

#### Context

Kubernetes has [Pod Security Standards](https://kubernetes.io/docs/concepts/security/pod-security-standards/). Mimir and GEM don't have technical requirements to run as anything other than the Restricted policy.

These policies can be enforced in two ways:
* via a [PodSecurityPolicy (PSP)](https://v1-24.docs.kubernetes.io/docs/concepts/security/pod-security-policy/), which was deprecated in Kubernetes 1.21 and removed in 1.25
* via [namespace labels](https://kubernetes.io/docs/tasks/configure-pod-container/enforce-standards-namespace-labels/), which was introduced in 1.22


Complying with the Restricted policy translates to running pods with the following SecurityContexts:

```yaml
  securityContext:
    fsGroup: 10001
    runAsGroup: 10001
    runAsNonRoot: true
    runAsUser: 10001
    seccompProfile:
      type: RuntimeDefault

  containerSecurityContext:
    allowPrivilegeEscalation: false
    readOnlyRootFilesystem: true
    capabilities:
      drop: [ALL]
```


#### What this PR does

This PR allows running most of the mimir-distributed chart under a Restricted policy. It adds these contexts to all GEM/Mimir components, nginx and memcached by default.

Running under Restricted policy is currently possible for minio and the Grafana Agent operator. I have two open PRs in the agent-operator chart that will enable users of mimir-distributed to configure the two charts in a way that they are compatible with the Restricted policy - https://github.com/grafana/helm-charts/pull/1829 and https://github.com/grafana/helm-charts/pull/1830. When those are merged and updated in the mimir-distributed chart, I will open another PR to allow metamonitoring to comply with the Restricted policy. Since minio is not recommended for production setups, I did not look into making minio compliant.

#### Notes to reviewers 


* The currently latest GEM image r204 doesn't support these settings. So this PR should be merged after r205 is released and used in the chart. Consequently, the CI also fails because the current r203 in the chart cannot be installed.

* In order to use the same SecurityContext with both namespace labels and PSP, I had to amend the PSP we have to allow
setting a seccomp profile to pods and to restrict it to the RuntimeDefault profile. This is already a requirement
when using namespace labels, so this change strengthens the existing PSP.



#### Testing procedure

I ran a few tests with these changes.

* On Kubernetes 1.21
  * Upgrading Mimir from before this PR to this PR. This verifies that the `fsGroup` change is automatic and sufficient, and that doesn't need manual intervention[^1].
  * Upgrading GEM with the same considerations as above.
  * Doing a fresh installation of Mimir and enabling PSP admission controller, so PSP is enforced.
  * Doing a fresh installation of GEM and enabling PSP admission controller, so PSP is enforced.
* On Kubernetes 1.25
  * Running with results cache.
  * Doing a fresh installation of Mimir and enforcing a Restricted policy.
  * Doing a fresh installation of GEM and enforcing a Restricted policy.

[^1]: This only works when the CSI driver supports fsGroup or when kubelet can do the permissions change itself. It does not work for HostPath persistent volumes, which is what most local dev distributions of K8s use. This shouldn't be the case in production setups.

If you want to reproduce the above, below are some tips:

<details>
<summary>tips</summary>

* GEM's r204 doesn't support running as Restricted because it has capabilities set. Use a Docker image from a recent master branch build or wait for r205 next week.
* Minio is not compliant with Restricted either. Disable minio and use a cloud storage provider. I used GCS.
* When running k8s <1.25 in order to have PSPs enforced you need to enable them. This is an article on how to do that in general. For k3d use this command when creating the cluster: `k3d cluster create helm --k3s-arg='--kube-apiserver-arg=enable-admission-plugins=PodSecurityPolicy@'`
* In k3d when running with PSP enabled some pods in the kube-system namespace will not start because they aren't using a PSP. I created an unrestricted PSP that applies to all of the `kube-system` namespace. This allowed the cluster to bootstrap correctly. See here for the YAML manifest of that PSP and some additional resources. I created everything after and including the `system-unrestricted-psp` resource from [this section in k3s' docs](https://rancher.com/docs/k3s/latest/en/security/hardening_guide/#pod-security-policies).

</details>

#### Which issue(s) this PR fixes or relates to

Fixes #2595

Fixes https://github.com/grafana/helm-charts/issues/1401

Fixes #2201 

#### Checklist

- NA Documentation added
